docs/proposals/control-data-plane-split/README.md

```diff
@@ -0,0 +1,265 @@
+- The communication channel between the control and data planes can be encrypted
+- Data planes can register with control plane
+- Data plane can scale independently of the control plane
+- RBAC policy for data plane follows the principle of least privilege. The data plane should not have access to the
+  Kubernetes API server.
+- RBAC policy for control plane follows the principle of least privilege.
+- Control plane provisions the data plane when a Gateway resource is created.
+
+## Design
+
+We will be using [NGINX Agent v3](https://github.com/nginx/agent/tree/v3) as our agent for the data plane. Our control plane and agent will connect to each other over a secure gRPC channel. The control plane will send nginx configuration updates over the channel using the agent's API, and the agent will write the files and reconfigure nginx.
+
+Whenever a user creates a Gateway resource, the control plane will provision an nginx deployment and service for that Gateway. The nginx/agent deployment will register itself with the control plane, and any Gateway API configuration that is created by a user for that Gateway will be sent to that nginx deployment.
+
+### Deployment Architecture
+
+![Deployment architecture](deployment-architecture.png)
+
+- _Control Plane Deployment_: The control plane is a Kubernetes Deployment with one container running the NGF
+  controller. The control plane will perform the same functions as it does today, but instead of
+  configuring nginx by writing files to a shared volume, it will send the configuration to the agent via gRPC.
+- _Control Plane Service_: Exposes the control plane via a Kubernetes Service of type `ClusterIP`. The data plane will
+  use the DNS name of the Service to connect to the control plane.
+- _Data Plane DaemonSet/Deployment_: The data plane can be deployed as either a DaemonSet or Deployment. The data
+  plane contains a single container running both the agent and nginx processes. The agent will download the
+  configuration from the control plane over a streaming RPC.
+- _NGINX Service_: Exposes nginx via a Kubernetes Service of type `LoadBalancer .` This is the entry point for the
+  customer’s traffic. Note that this Service should not expose any of the agent’s ports.
+
+#### Further Requirements and Implementation Details
+
+- Both deployments should have read only filesystems.
+- Both deployments should have the minimal permissions required to perform their functions.
+- The nginx deployment should be configurable via the helm chart.
+  - Downside of this is that these options will apply to all nginx instances.
+  - We could introduce a CRD, but where would it attach? We already have NginxProxy which controls dynamic data plane configuration, and this may eventually attach to the Gateway instead of just the GatewayClass. Would a Deployment configuration fit in there, and would it be dynamic? That would require us to completely redeploy nginx if a user changes those settings.
```

Comment:
It would make sense to attach this to the Gateway now since each Gateway is essentially one instance of NGINX (that could be horizontally scaled). But I think this is also best reserved for when we have demand for it.

Reply:
The main worry I have is that NginxProxy could eventually be updated to attach to Gateway instead of just GatewayClass. I could see that as valuable to a user.

But we can only attach one thing to a Gateway. So if the nginx Deployment config is in a separate CRD, a user couldn't configure that AND configure NginxProxy for specific nginx conf settings.

I think this is something we figure out in the future if there is demand for it. For now, the settings are all created on installation via a ConfigMap and applies to all nginx instances.